src/wallet/wallet.cpp

```diff
@@ -555,6 +558,124 @@ void CWallet::UpgradeDescriptorCache()
     SetWalletFlag(WALLET_FLAG_LAST_HARDENED_XPUB_CACHED);
 }
 
+/**
+ * UpgradeToGlobalHDKey searches through the descriptors in a descriptor wallet and tries to find a CExtPubKey
+ * which is likely to be the most recent one used to generate all of the automatically generated descriptors.
+ * The automatically generated descriptors are guaranteed to be 2 p2pkh, 2 p2sh-segwit, and 2 p2wpkh. There may
+ * be 2 p2tr descriptors, as well as imports. Those are ignored as they are not guaranteed to be present in all wallets.
+ * Candidate CExtKeys are those that appear in the aforementioned descriptors as the only key. The best candidate is
+ * selected as the one that was most recently used.
+ *
+ * The best candidate will be added to the wallet as the current active HD key. Any other candidates will be added as
+ * HD keys that have been rotated out.
+ *
+ * WALLET_FLAG_GLOBAL_HD_KEY will be set in order to indicate that this upgrade has occurred.
+ */
+void CWallet::UpgradeToGlobalHDKey(const std::map<DescIDKeyIDPair, CKey>& desc_keys, const std::map<DescIDKeyIDPair, CryptedKeyPair>& desc_crypt_keys)
+{
+    if (!IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS) ||
+        IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) ||
+        IsWalletFlagSet(WALLET_FLAG_GLOBAL_HD_KEY)
+        ) {
+        return;
+    }
```

Comment:
In 3e250b10:

There is a discrepancy here. The upgrade mechanism is skipped for private key disabled wallets without set the `WALLET_FLAG_GLOBAL_HD_KEY`. This differs from what is being done in the wallet creation and the migration processes where we set this flag for all types (only requirement is `WALLET_FLAG_DESCRIPTORS`).

Is this intended?

Reply:
I don't think it particularly matters since a wallet without private keys cannot have a global hd key. However I have changed it so that it is set only if private keys are enabled.